keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentPromotionChecker.kt

```diff
@@ -111,6 +118,15 @@ class EnvironmentPromotionChecker(
     }
   }
 
+  /**
+   * @return `true` if this artifact is used by any resource in [environment], `false` otherwise.
+   */
+  private fun DeliveryArtifact.isUsedIn(environment: Environment) =
+    environment
+      .resources
+      .map { (it.spec as? ArtifactReferenceProvider)?.artifactReference }
+      .contains(reference)
```

Comment:
This is basically the logic from the old `ArtifactUsedConstraintEvaluator` but it's not necessary to check it for each version as the result will never vary by version